Fix Share extension duplicates

thisIsTheFoxe opened the pull request:
1. move all files into to the app
2. check if file with same path exists
finally, ask user to resolve conflicts.
- Use `DispatchQueue` to keep track of all (un-)uploaded files. When done, dismiss.
- Don't let user rename with already existing filename, feedback welcome:
  -  just return and do nothing (current implementation of the PR)
  - show an error
  - disable rename button, if text matches one of the other files

TODO:
- [x] Test
- [ ] Error handling

marinofaggiana commented:
@thisIsTheFoxe you are completely broken the code ... how I can check it out ? it was necessary to have to rewrite it ?

thisIsTheFoxe commented:
@marinofaggiana 
1. done (see discussion in the 2nd issue)
2. done, the error is still needed when renaming a file since you can't have the same filename twice
4. I've tested it with a bunch of files and seems to work fine

marinofaggiana commented:
> @marinofaggiana
> 
> 1. done (see discussion in the 2nd issue)

I don't look the "..." for 1 file ?

<img width="343" alt="Schermata 2022-01-08 alle 11 21 21" src="https://user-images.githubusercontent.com/8616947/148640684-a30d7ac0-e55d-42ed-ab1b-bd50d538fa0a.png">
<img width="343" alt="Schermata 2022-01-08 alle 11 21 06" src="https://user-images.githubusercontent.com/8616947/148640686-12562dc2-5a30-420e-a97c-cea41d8b9ba5.png">

> 3. I've tested it with a bunch of files and seems to work fine

Good !

- new issue, counter error: Upload file 0/1 ?

thisIsTheFoxe commented:
> I don't look the symbol rename for 1 file ?

The idea was that a person cannot remove the last file, since then there'd be nothing left to upload. But to not overcomplicate things I reverted it as it was before. 

> new issue, counter error: Upload file 0/1 ?

Seemed fine to me, 0 / 1 file is uploaded, so it's uploading the first file. But I see how that could be confusing.. fixed it.

marinofaggiana commented:
> > I don't look the symbol rename for 1 file ?
> 
> The idea was that a person cannot remove the last file, since then there'd be nothing left to upload. But to not overcomplicate things I reverted it as it was before.

Eventually we can consider replacing (for only 1) the "..." with the "pencil" to change only the name? what do you think ?

thisIsTheFoxe commented:
Not quite sure what you mean by "eventually".. why would we postpone it?

But, I also considered that. And that was why I said _not to overcomplicate things_. It's an idea but I don't really think it's worth it..

marinofaggiana commented:
Ok, can you rebase this PR so I can test it with my HW ?

thisIsTheFoxe commented:
Done.

marinofaggiana commented:
@thisIsTheFoxe I have try with my iPhone 12 Pro Max iOS 15.2, two image, one of 5 MB and one of 500 kb, when detect that the file exists crashed ...

Thread 1: EXC_RESOURCE RESOURCE_TYPE_MEMORY (limit=120 MB, unused=0x0)

as I imagined ...

thisIsTheFoxe commented:
Can confirm 👍 
I tried it with like 20+ images but I had none that was bigger than 5MB :D

Anyway, downsampling them before displaying solves it. Tried it with a 8k (17MB) with a max memory usage of like 20MB.

marinofaggiana commented:
Good, seems now works but still some problems (I don't know if it was before but it doesn't matter).
Upload several file with different size, example 10 file from 5 kb until video over 100 MB (chunk) ...  the IHProgressHUD seems not to work well and if I cancel the share during the upload the files remain in queue ( try and open the app ) I don't think it's correct ... is better use the Upload without the background (??? this is dangerous),  what's do you think ? 

if you don't know how to do it, I'll gladly help you.

marinofaggiana commented:
Note. I checked the code .. in fact it didn't, it was simpler but more functional, better to send the files one by one; as I told you before making changes ask me, many things always have a certain logic behind them. please ask! 
Doing so yes but half a step forward and 4 back.

marinofaggiana commented:
+ Added bug: Screen capture of iPhone and send to Share ... missing the extension.

thisIsTheFoxe commented:
Hmm.. for the record, I still think there's a big benefit in parallelizing here, especially if you have many smaller files.. it takes very long..
But I also see how not being able to cancel request is quite a drawback.. I'll revert..

> Screen capture of iPhone and send to Share ... missing the extension

Not sure what you mean. Screenshots as well as screen recordings (saved as video in Photos) can be shared without problems for me.

marinofaggiana commented:
Good, now works but remain the upload in progress, if the user close the share the upload remains ... and it is the case to eliminate it .... so you can detect when the extension is closed and remove the metadata  before the exit.

line 358 let metadata = uploadMetadata.removeFirst()

thisIsTheFoxe commented:
@marinofaggiana I'm not sure what you mean. I use `uploadStarted` to ensure that the extension is active. If the user cancels it, no further uploads are started..

marinofaggiana commented:
The problem is the following:

The user selects File n where n can only be one, he realizes that the operation is wrong (he is sending the file to an incorrect location or the file is wrong, for example it is huge) by forcibly closing the share all files must be deleted and the uploaded file must be stopped and deleted.
Now the upload file remains in the queue and when you open the Nextcloud application it resumes.

I don't know if I'm clear now, try it yourself, the important thing is that no file remains pending, if I close everything it must be deleted.

marinofaggiana commented:
@thisIsTheFoxe do you have done some experiment to see if it is possible to intercept the dismiss of UIViewController without the cancel button (Swipe down the view)? (but ... I don't think it can be)

marinofaggiana reviewed (commented):
Seems works.
